pyro/infer/mcmc/hmc.py

```diff
@@ -0,0 +1,225 @@
+from pyro.infer import TracePosterior
+from pyro.util import ng_ones, ng_zeros
+from tests.common import assert_equal
+
+
+def verlet_integrator(z, r, grad_potential, step_size, num_steps):
+    """
+    Velocity Verlet integrator.
+
+    :param z: dictionary of sample site names and their current values
+    :param r: dictionary of sample site names and corresponding momenta
+    :param grad_potential: function that returns gradient of the potential given z
+        for each sample site
+    :return: (z_next, r_next) having same types as (z, r)
+    """
+    # deep copy the current state - (z, r)
+    z_next = {key: val.clone().detach() for key, val in z.items()}
+    r_next = {key: val.clone().detach() for key, val in r.items()}
+    retain_grads(z_next)
+    retain_grads(r_next)
+    grads = grad_potential(z_next)
+
+    for _ in range(num_steps):
+        # detach graph nodes for next iteration
+        detach_nodes(z_next)
+        detach_nodes(r_next)
+        for site_name in z_next:
+            # r(n+1/2)
+            r_next[site_name] = r_next[site_name] + 0.5 * step_size * (-grads[site_name])
+            # z(n+1)
+            z_next[site_name] = z_next[site_name] + step_size * r_next[site_name]
+        # retain gradients for intermediate nodes in backward step
+        retain_grads(z_next)
+        retain_grads(r_next)
+        grads = grad_potential(z_next)
+        for site_name in r_next:
+            # r(n+1)
+            r_next[site_name] = r_next[site_name] + 0.5 * step_size * (-grads[site_name])
+    return z_next, r_next
+
+
+def retain_grads(z):
+    for key in z:
+        # XXX: can be removed with PyTorch 0.3
+        if z[key].is_leaf and not z[key].requires_grad:
+            z[key].requires_grad = True
+        z[key].retain_grad()
+
+
+def detach_nodes(z):
+    for key in z:
+        z[key] = Variable(z[key].data, requires_grad=True)
+
+
+class HMC(TracePosterior):
+    def __init__(self,
+                 model,
+                 step_size=0.5,
+                 num_steps=3,
+                 warmup_steps=20,
+                 num_samples=1000):
+        self.model = model
+        self.step_size = step_size
+        self.num_steps = num_steps
+        self.num_samples = num_samples
+        self.warmup_steps = warmup_steps
+        # simulation run attributes - will be set in self._setup
+        # at start of run
+        self._cached_param_grads = {}
+        self._args = None
+        self._kwargs = None
+        self._accept_cnt = None
+        self._prototype_trace = None
+        super(HMC, self).__init__()
+
+    def _log_prob(self, z):
+        """
+        Return log pdf of the model with sample sites replayed from z_trace
+        """
+        z_trace = self._prototype_trace.copy()
+        for name, value in z.items():
+            z_trace.nodes[name]['value'] = value
+        model_trace = poutine.trace(poutine.replay(self.model, trace=z_trace)) \
```

Comment:
I'm pretty sure you'll want a poutine here to `.detach()` params during replay so they won't accumulate gradients. What's implemented now smells bad due to:
- extra work performed during `.backward()` to propagate gradients all the way back to params, and
- dirty `.grad`s on all params in the param store (and it also smells bad to `zero_grads()` on the param store).

One way to get this behavior is to change `poutine.replay()` to replay `param` sites (in addition to `sample` sites). That way you can `.detach()` all param sites in the prototype trace and get the desired behavior. @eb8680 Do you have any objections to `poutine.replay()` behaving this way, so as to avoid yet another poutine?

Reply:
(or does `poutine.replay()` already do the right thing?)

Reply:
I can override `_pyro_param` in ReplayPoutine and reinstate some of the behavior that I had to begin with, if we all agree with this modification to ReplayPoutine. :)

> extra work performed during .backward() to propagate gradients all the way back to params, and

Detaching will not prevent this extra gradient flow back to the params right?

> dirty .grads on all params in the param store (and it also smells bad to zero_grads() on the param store).

Currently this is handled by the `_cleanup` operation, so that gradients on the params are reset; but I completely agree that detaching would be the cleaner way of doing this.

Reply:
Sorry I can't respond to these very quickly, I'm busy working on a workshop submission due tomorrow, but I'm not sure about this `replay` change.

> I'm pretty sure you'll want a poutine here to .detach() params during replay so they won't accumulate gradients

This seems like premature optimization, but if you really want to account for this, just iterate over the parameters in `model_trace` and set `parameter.requires_grad=False` for each one immediately before you call `backward()`, then reset `parameter.requires_grad=True` immediately after.

Reply:
I think you'll want to do something like:
```py
for node in prototype_trace.nodes.values():
    if node['type'] == 'param':
        node['value'] = Variable(node['value'].data)  # detach and set requires_grad = False
    elif node['type'] == 'sample' and not node['is_observed']:
        node['value'] = Variable(node['value'].data, requires_grad=True)
```
which should indeed prevent gradients from being propagated to params.

Reply:
@fritzo - There is a subtle issue with this. Since the parameters are installed in site args/kwargs and registered with the param store, this will still propagate the gradients back to the param store.